### PR TITLE
Feature/more aliasing tests

### PR DESCRIPTION
**Pull request by serge-sans-paille**

### CODE REVIEW

**Comment by serge-sans-paille** [comment]
@pbrunet : 49-3 article has been used :-/

File: doc/INTERNAL.rst [hunk 1]
```diff
@@ -246,7 +246,7 @@ about aliased values::
     >>> foo_tree = getast(foo)
     >>> lazyness = pm.gather(analyses.LazynessAnalysis, foo_tree)
     >>> lazyness
-    {'a': 2, 'array': 1, 'b': 1, 'a_': 1}
+    {'a': 1, 'array': 1, 'b': 1, 'a_': 1}
```
Comment (pbrunet) [inline doc/INTERNAL.rst]: Ho?

File: pythran/analyses/aliases.py [hunk 2]
```diff
@@ -106,55 +95,164 @@ def rec(w, n):
...
+        >>> Aliases.dump(result, filter=ast.BoolOp)
+        (a or b) => ['a', 'b']
+
+        Note that a literal does not create any alias
+
+        >>> module = ast.parse('def foo(a, b): return a or 0')
+        >>> result = pm.gather(Aliases, module)
+        >>> Aliases.dump(result, filter=ast.BoolOp)
+        (a or 0) => ['<unbound-value>', 'a']
+        '''
```
Comment (pbrunet) [inline pythran/analyses/aliases.py]: So nice !!

File: pythran/analyses/aliases.py [hunk 3]
```diff
@@ -106,55 +95,164 @@ def rec(w, n):
...
+        <function merge_return_aliases at...>
+
+        This field is a function that takes as many nodes as the function
+        argument count as input and returns an expression based on
+        these arguments if the function happens to create aliasing
+        between its input and output. In our case:
+
+        >>> f = module.body[0].return_alias
+        >>> Aliases.dump(f([ast.Name('A', ast.Load()), ast.Num(1)]))
+        ['A']
```
Comment (pbrunet) [inline pythran/analyses/aliases.py]: Why not having 'UnboundValue" too here?

File: pythran/analyses/aliases.py [hunk 4]
```diff
@@ -292,15 +499,15 @@ def visit_DictComp(self, node):
     # aliasing created by statements
 
     def visit_FunctionDef(self, node):
-        """
+        '''
```
Comment (pbrunet) [inline pythran/analyses/aliases.py]: For information, PEP257 say we should use """ and not ''' for docstring.

File: pythran/analyses/aliases.py [hunk 5]
```diff
@@ -363,38 +569,118 @@ def merge_return_aliases(args):
...
+
+        >>> from pythran import passmanager
+        >>> pm = passmanager.PassManager('demo')
+        >>> fun = """
+        ... def foo(a):
+        ...     while(a):
+        ...         if a==1: print b
+        ...         else: b=a"""
+        >>> module = ast.parse(fun)
+        >>> result = pm.gather(Aliases, module)
```
Comment (pbrunet) [inline pythran/analyses/aliases.py]: No check for values?

File: pythran/backend.py [hunk 6]
```diff
@@ -28,12 +28,7 @@
 from math import isnan, isinf
 import ast
 import os
-import sys
-
-if sys.version_info[0] < 3:
-    import cStringIO
-else:
-    from io import StringIO as cStringIO
+import StringIO
```
Comment (pbrunet) [inline pythran/backend.py]: :) finally no more Python3 support :)

File: pythran/pythonic/__builtin__/float_.hpp [hunk 7]
```diff
@@ -34,6 +34,11 @@ namespace pythonic
       {
         return mpz_get_d(a.get_mpz_t());
       }
+      template <class T, class U>
+      float_::type float_::operator()(__gmp_expr<T, U> &a)
+      {
+        return mpz_get_d(a.get_mpz_t());
+      }
```
Comment (pbrunet) [inline pythran/pythonic/__builtin__/float_.hpp]: why?

File: pythran/pythonic/include/__builtin__/set/add.hpp [hunk 8]
```diff
@@ -14,10 +15,13 @@ namespace pythonic
...
       template <class T, class F>
-      void add(types::set<T> &s, F const &value);
+      types::none_type add(types::set<T> &s, F const &value);
 
       template <class T, class F>
-      void add(types::set<T> &&s, F const &value);
+      types::none_type add(types::set<T> &&s, F const &value);
+
+      template <class F>
+      types::none_type add(types::empty_set const &s, F &&value);
```
Comment (pbrunet) [inline pythran/pythonic/include/__builtin__/set/add.hpp]: This one is declared but never defined?

File: doc/INTERNAL.rst (hunk 1 above)
Comment (serge-sans-paille) [inline doc/INTERNAL.rst]: Yeah, Icannot tell you why. if you can investigate this one...

File: pythran/analyses/aliases.py (hunk 3 above)
Comment (serge-sans-paille) [inline pythran/analyses/aliases.py]: Why would you want one? You know for sure you'll alias  to ``a``, if I'm not mistaken.

File: pythran/analyses/aliases.py (hunk 4 above)
Comment (serge-sans-paille) [inline pythran/analyses/aliases.py]: ok ;-)

File: pythran/analyses/aliases.py (hunk 5 above)
Comment (serge-sans-paille) [inline pythran/analyses/aliases.py]: No, I don't know what to check here... maybe return b and check the return_alias?

File: pythran/backend.py (hunk 6 above)
Comment (serge-sans-paille) [inline pythran/backend.py]: ooops, I did not meant that :-/

File: pythran/pythonic/__builtin__/float_.hpp (hunk 7 above)
Comment (serge-sans-paille) [inline pythran/pythonic/__builtin__/float_.hpp]: API update of GMP

File: pythran/pythonic/include/__builtin__/set/add.hpp (hunk 8 above)
Comment (serge-sans-paille) [inline pythran/pythonic/include/__builtin__/set/add.hpp]: yep, it's only there for type inference in __combined

**Pull request merged**